### PR TITLE
Recover from EINVAL even if the error is untyped

### PR DESCRIPTION
**Pull request by cyberdelia**
Recover from EINVAL even if the error is untyped.

Swift version: 
```
Apple Swift version 5.2.4 (swiftlang-1103.0.32.9 clang-1103.0.32.53)
Target: x86_64-apple-darwin19.6.0
```
NIO version: `2.19.0`

### Motivation:

I'm facing an issue an EINVAL error is raised and the whole pipeline is shutdown silently. The EINVAL error is related to `fnctl` call as well.
The circumstances in which this appears are still unclear, and I'm not familiar enough with the codebase to make a call. 

### Modifications:

This just ignore EINVAL error code as an error.

### Result:

This prevent the pipeline from shutting down silently under condition where it clearly has no reasons too.

### CODE REVIEW

**Comment by Lukasa** [comment]
Hmm, I'm reluctant to merge this without more information.

Per the manual pages from both macOS and Linux, `EINVAL` on `accept` implies one of the following conditions:

1. Socket cannot accept connections
2. `addrlen` is invalid
3. `accept4` has been passed invalid flags.

None of these cases are recoverable: we'd expect to see them all continue to trigger. In those cases, this change would turn this into a 100% CPU spin loop, which is not good.

I think we need to know what's going on in your system to understand what the actual problem is.

**Comment by cyberdelia** [comment]
@Lukasa Agreed.

But, so far haven't been able to point my finger on why EINVAL is returned or come up with a small scale reproduction case. Essentially, the same Encoder is either returning a response that triggers the error or not, which is very confusing.

**Comment by Lukasa** [comment]
On what platforms do you see the EINVAL return value?

**Comment by cyberdelia** [comment]
@Lukasa macOS (latest stable release).

**Comment by Lukasa** [comment]
Do you see reproduction with any consistency?

**Comment by cyberdelia** [comment]
@Lukasa I can reproduce easily, but I have been able to pinpoint what is triggering the behaviour.

**Comment by Lukasa** [comment]
Ok, do you want to add a breakpoint on `read` that will fire conditionally if `errno` is `EINVAL`? It'd be useful to see the state of the system when that fires.

**Comment by cyberdelia** [comment]
Can do, will report later!

**Comment by cyberdelia** [comment]
I've added a breakpoint `shouldCloseOnReadError`, here is the backtrace when encountering the error:

```
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
  * frame #0: 0x000000010018f0a9 agent`ServerSocketChannel.shouldCloseOnReadError(err=NIO.IOError @ 0x000060000260c1c8, self=0x0000000100d04400) at SocketChannel.swift:259:20
    frame #1: 0x0000000100041bcd agent`BaseSocketChannel.readable0(self=0x0000000100d04400) at BaseSocketChannel.swift:1092:21
    frame #2: 0x0000000100042c82 agent`BaseSocketChannel.readable(self=0x0000000100d04400) at BaseSocketChannel.swift:1044:14
    frame #3: 0x0000000100044109 agent`protocol witness for SelectableChannel.readable() in conformance BaseSocketChannel<A> at <compiler-generated>:0
    frame #4: 0x0000000100159e2a agent`SelectableEventLoop.handleEvent<C>(ev=(rawValue = 4), channel=0x0000000100d04400, self=0x0000600003b0c000) at SelectableEventLoop.swift:323:25
    frame #5: 0x000000010015b934 agent`closure #1 in closure #2 in SelectableEventLoop.run(ev=NIO.SelectorEvent<NIO.NIORegistration> @ 0x00007ffeefbfc9c8, self=0x0000600003b0c000) at SelectableEventLoop.swift:398:30
    frame #6: 0x000000010015bd39 agent`thunk for @callee_guaranteed (@guaranteed SelectorEvent<NIORegistration>) -> (@error @owned Error) at <compiler-generated>:0
    frame #7: 0x0000000100160374 agent`partial apply for thunk for @callee_guaranteed (@guaranteed SelectorEvent<NIORegistration>) -> (@error @owned Error) at <compiler-generated>:0
    frame #8: 0x0000000100170974 agent`Selector.whenReady(strategy=block, body=0x0000000100160360 agent`partial apply forwarder for reabstraction thunk helper from @callee_guaranteed (@guaranteed NIO.SelectorEvent<NIO.NIORegistration>) -> (@error @owned Swift.Error) to @escaping @callee_guaranteed (@in_guaranteed NIO.SelectorEvent<NIO.NIORegistration>) -> (@error @owned Swift.Error) at <compiler-generated>, self=0x0000600002108140) at Selector.swift:561:17
    frame #9: 0x000000010015b6c9 agent`closure #2 in SelectableEventLoop.run(self=0x0000600003b0c000, nextReadyTask=nil) at SelectableEventLoop.swift:395:36
    frame #10: 0x000000010015f2c4 agent`partial apply for closure #2 in SelectableEventLoop.run() at <compiler-generated>:0
    frame #11: 0x000000010003101f agent`thunk for @callee_guaranteed () -> (@error @owned Error) at <compiler-generated>:0
    frame #12: 0x000000010015f2e4 agent`thunk for @callee_guaranteed () -> (@error @owned Error)partial apply at <compiler-generated>:0
    frame #13: 0x00000001001560c2 agent`closure #1 in withAutoReleasePool<T>(execute=0x000000010015f2d0 agent`reabstraction thunk helper from @callee_guaranteed () -> (@error @owned Swift.Error) to @escaping @callee_guaranteed () -> (@out (), @error @owned Swift.Error)partial apply forwarder with unmangled suffix ".8" at <compiler-generated>) at SelectableEventLoop.swift:23:13
    frame #14: 0x000000010015611f agent`partial apply for closure #1 in withAutoReleasePool<A>(_:) at <compiler-generated>:0
    frame #15: 0x00007fff73219d8e libswiftObjectiveC.dylib`ObjectiveC.autoreleasepool<A>(invoking: () throws -> A) throws -> A + 46
    frame #16: 0x0000000100156069 agent`withAutoReleasePool<T>(execute=0x000000010015f2d0 agent`reabstraction thunk helper from @callee_guaranteed () -> (@error @owned Swift.Error) to @escaping @callee_guaranteed () -> (@out (), @error @owned Swift.Error)partial apply forwarder with unmangled suffix ".8" at <compiler-generated>) at SelectableEventLoop.swift:22:16
    frame #17: 0x000000010015a380 agent`SelectableEventLoop.run(self=0x0000600003b0c000) at SelectableEventLoop.swift:394:17
    frame #18: 0x00000001000d9d73 agent`static MultiThreadedEventLoopGroup.runTheLoop(thread=0x0000600000c080c0, canEventLoopBeShutdownIndividually=true, selectorFactory=0x00000001000e1670 agent`partial apply forwarder for NIO.Selector.__allocating_init() throws -> NIO.Selector<A> at <compiler-generated>, initializer=0x00000001000ddc50 agent`closure #1 (NIO.NIOThread) -> () in static NIO.MultiThreadedEventLoopGroup.withCurrentThreadAsEventLoop((NIO.EventLoop) -> ()) -> () at EventLoop.swift:1027, callback=0x00000001000df5d0 agent`partial apply forwarder for closure #2 (NIO.SelectableEventLoop) -> () in static NIO.MultiThreadedEventLoopGroup.withCurrentThreadAsEventLoop((NIO.EventLoop) -> ()) -> () at <compiler-generated>, self=NIO.MultiThreadedEventLoopGroup) at EventLoop.swift:824:22
    frame #19: 0x00000001000ddc20 agent`static MultiThreadedEventLoopGroup.withCurrentThreadAsEventLoop(callback=0x000000010001c1a0 agent`partial apply forwarder for closure #1 (NIO.EventLoop) -> () in static Agent.Daemon.start(address: Swift.String) throws -> () at <compiler-generated>, self=NIO.MultiThreadedEventLoopGroup) at EventLoop.swift:1024:37
```

And variables:

```
(NIO.IOError) err = {
  failureDescription = "fcntl(descriptor:command:value:)"
  error = errno (errno = 22)
}
(NIO.ServerSocketChannel) self = 0x0000000100d04400 {
  NIO.BaseSocketChannel = {
    parent = nil
    socket = 0x00006000002088c0 {
      NIO.BaseSocket = {
        descriptor = 4
      }
    }
    closePromise = {
      futureResult = 0x00006000021082d0 {
        _value = nil
        eventLoop = 0x0000600003b0c000 {
          _selector = 0x0000600002108140 {
            lifecycleState = open
            eventsCapacity = 64
            events = 0x0000000102810200
            registrations = 1 key/value pair {
              [0] = {
                key = 4
                value = serverSocketChannel {
                  serverSocketChannel = {
                    0 = 0x0000000100d04400 {...}
                    1 = (rawValue = 7)
                  }
                }
              }
            }
            deregistrationsHappened = false
            externalSelectorFDLock = 0x0000600000208280 (mutex = 0x0000000100d04080)
            selectorFD = 3
            myThread = 0x0000600000c08150 {
              desiredName = nil
              handle = 0x00000001008dbdc0
            }
          }
          thread = 0x0000600000c080c0 {
            desiredName = nil
            handle = 0x00000001008dbdc0
          }
          _scheduledTasks = {
            _heap = {
              type = minHeap
              storage = {
                _buffer = {
                  _storage = 0 values {}
                }
              }
              comparator = 0x000000010010b790 agent`partial apply forwarder for dispatch thunk of static Swift.Comparable.< infix(A, A) -> Swift.Bool at <compiler-generated>
            }
          }
          tasksCopy = {
            _buffer = {
              _storage = 0 values {}
            }
          }
          canBeShutdownIndividually = true
          _tasksLock = 0x00006000002082e0 (mutex = 0x0000000100d040c0)
          _externalStateLock = 0x0000600000208300 (mutex = 0x0000000100d04100)
          internalState = runningAndAcceptingNewRegistrations
          externalState = open
          _iovecs = 0x0000000102810a00
          _storageRefs = 0x0000000102814a00
          iovecs = (_position = 0x0000000102810a00, count = 1024)
          storageRefs = (_position = 0x0000000102814a00, count = 1024)
          _msgs = 0x0000000100aae000
          _addresses = 0x0000000100abd000
          msgs = (_position = 0x0000000100aae000, count = 1024)
          addresses = (_position = 0x0000000100abd000, count = 1024)
          promiseCreationStoreLock = 0x0000600000208320 (mutex = 0x0000000100d04140)
          _promiseCreationStore = 1 key/value pair {
            [0] = {
              key = 105556287506573
              value = (file = "~/Library/Developer/Xcode/DerivedData/agent-bzyktklnmtouupfkehbiuetsgxph/SourcePackages/checkouts/swift-nio/Sources/NIO/BaseSocketChannel.swift", line = 445)
            }
          }
        }
        _callbacks = {
          firstCallback = nil
          furtherCallbacks = nil
        }
      }
    }
    selectableEventLoop = 0x0000600003b0c000 {...}
    _offEventLoopLock = 0x0000600000208920 (mutex = 0x0000000100d04180)
    isActiveAtomic = 0x0000600000208980 {}
    socketDescription = "BaseSocket { fd=4 }"
    readPending = true
    pendingConnect = nil
    recvAllocator = {
      minIndex = 3
      maxIndex = 38
      minimum = 64
      maximum = 65536
      initial = 1024
      index = 32
      nextReceiveBufferSize = 1024
      decreaseNow = false
    }
    maxMessagesPerRead = 4
    inFlushNow = false
    autoRead = true
    _pipeline = 0x0000600002108320 {
      head = 0x0000600002608300 {
        next = 0x00006000026000c0 {
          next = 0x0000600002608360 {
            next = nil
            prev = 0x00006000026000c0 {...}
            pipeline = 0x0000600002108320 {...}
            name = "tail"
            inboundHandler = 0x000060000000c010
            outboundHandler = nil
            removeHandlerInvoked = false
            userTriggeredRemovalStarted = false
          }
          prev = 0x0000600002608300 {...}
          pipeline = 0x0000600002108320 {...}
          name = "AcceptHandler"
          inboundHandler = 0x0000600000c00390 {
            childChannelInit = 0x000000010001be30 agent`closure #1 (NIO.Channel) -> NIO.EventLoopFuture<()> in static Agent.Daemon.(daemonBootstrap in _5158C76E4BA214CC96ED6AC320AE28CC)(group: NIO.EventLoopGroup) -> NIO.ServerBootstrap at Server.swift:285
            childChannelOptions = {
              _storage = 0 values {}
            }
          }
          outboundHandler = nil
          removeHandlerInvoked = false
          userTriggeredRemovalStarted = false
        }
        prev = nil
        pipeline = 0x0000600002108320 {...}
        name = "head"
        inboundHandler = nil
        outboundHandler = 0x000060000000c000
        removeHandlerInvoked = false
        userTriggeredRemovalStarted = false
      }
      tail = 0x0000600002608360 {...}
      idx = 0
      destroyed = false
      eventLoop = 0x0000600003b0c000 {...}
      _channel = 0x0000000100d04400 {...}
    }
    _addressCache = {
      local = unixDomainSocket {
        unixDomainSocket = {
          _storage = 0x8000600002c00000 {
            value = {
              sun_len = 106
              sun_family = 1
              sun_path = {
                0 = 47
                1 = 118
                2 = 97
                3 = 114
                4 = 47
                5 = 102
                6 = 111
                7 = 108
                8 = 100
                9 = 101
                10 = 114
                11 = 115
                12 = 47
                13 = 104
                14 = 99
                15 = 47
                16 = 113
                17 = 54
                18 = 54
                19 = 118
                20 = 116
                21 = 95
                22 = 56
                23 = 53
                24 = 55
                25 = 95
                26 = 108
                27 = 50
                28 = 49
                29 = 113
                30 = 113
                31 = 107
                32 = 48
                33 = 108
                34 = 121
                35 = 49
                36 = 55
                37 = 120
                38 = 102
                39 = 56
                40 = 48
                41 = 48
                42 = 48
                43 = 48
                44 = 103
                45 = 110
                46 = 47
                47 = 84
                48 = 47
                49 = 97
                50 = 103
                51 = 101
                52 = 110
                53 = 116
                54 = 46
                55 = 49
                56 = 56
                57 = 50
                58 = 48
                59 = 51
                60 = 0
                61 = 0
                62 = 0
                63 = 0
                64 = 0
                65 = 0
                66 = 0
                67 = 0
                68 = 0
                69 = 0
                70 = 0
                71 = 0
                72 = 0
                73 = 0
                74 = 0
                75 = 0
                76 = 0
                77 = 0
                78 = 0
                79 = 0
                80 = 0
                81 = 0
                82 = 0
                83 = 0
                84 = 0
                85 = 0
                86 = 0
                87 = 0
                88 = 0
                89 = 0
                90 = 0
                91 = 0
                92 = 0
                93 = 0
                94 = 0
                95 = 0
                96 = 0
                97 = 0
                98 = 0
                99 = 0
                100 = 0
                101 = 0
                102 = 0
                103 = 0
              }
            }
          }
        }
      }
      remote = nil
    }
    _bufferAllocatorCache = (malloc = 0x0000000100067fc0 agent`@objc closure #1 (Swift.Int) -> Swift.Optional<Swift.UnsafeMutableRawPointer> in NIO.ByteBufferAllocator.init() -> NIO.ByteBufferAllocator at <compiler-generated>, realloc = 0x0000000100068010 agent`@objc closure #2 (Swift.Optional<Swift.UnsafeMutableRawPointer>, Swift.Int) -> Swift.Optional<Swift.UnsafeMutableRawPointer> in NIO.ByteBufferAllocator.init() -> NIO.ByteBufferAllocator at <compiler-generated>, free = 0x0000000100068050 agent`@objc closure #3 (Swift.Optional<Swift.UnsafeMutableRawPointer>) -> () in NIO.ByteBufferAllocator.init() -> NIO.ByteBufferAllocator at <compiler-generated>, memcpy = 0x00000001000680b0 agent`@objc closure #4 (Swift.UnsafeMutableRawPointer, Swift.UnsafeRawPointer, Swift.Int) -> () in NIO.ByteBufferAllocator.init() -> NIO.ByteBufferAllocator at <compiler-generated>)
    interestedEvent = (rawValue = 7)
    lifecycleManager = {
      eventLoop = 0x0000600003b0c000 {...}
      isActiveAtomic = 0x0000600000208980 {...}
      hasSeenEOFNotification = false
      currentState = activated
    }
    bufferAllocator = (malloc = 0x0000000100067fc0 agent`@objc closure #1 (Swift.Int) -> Swift.Optional<Swift.UnsafeMutableRawPointer> in NIO.ByteBufferAllocator.init() -> NIO.ByteBufferAllocator at <compiler-generated>, realloc = 0x0000000100068010 agent`@objc closure #2 (Swift.Optional<Swift.UnsafeMutableRawPointer>, Swift.Int) -> Swift.Optional<Swift.UnsafeMutableRawPointer> in NIO.ByteBufferAllocator.init() -> NIO.ByteBufferAllocator at <compiler-generated>, free = 0x0000000100068050 agent`@objc closure #3 (Swift.Optional<Swift.UnsafeMutableRawPointer>) -> () in NIO.ByteBufferAllocator.init() -> NIO.ByteBufferAllocator at <compiler-generated>, memcpy = 0x00000001000680b0 agent`@objc closure #4 (Swift.UnsafeMutableRawPointer, Swift.UnsafeRawPointer, Swift.Int) -> () in NIO.ByteBufferAllocator.init() -> NIO.ByteBufferAllocator at <compiler-generated>)
  }
  backlog = 128
  group = 0x0000600003b0c000 {...}
}
(NIO.IOError) err = {
  failureDescription = "fcntl(descriptor:command:value:)"
  error = errno (errno = 22)
}
```

**Comment by Lukasa** [comment]
Interesting. The error is an `IOError` from `fcntl`. We defend against this today when we set the socket non-blocking: https://github.com/apple/swift-nio/blob/08706eb13e063a82542fb047633140f99a2e7208/Sources/NIO/System.swift#L585-L597

However, there's another `fcntl` in the `accept` path, very low down the stack: https://github.com/apple/swift-nio/blob/08706eb13e063a82542fb047633140f99a2e7208/Sources/NIO/System.swift#L298-L323

I don't see any reason to think that this `fcntl` is any different than the other one: in principle we can hit `EINVAL` here just as much as we can when we try to set the socket non-blocking.

This call to `fcntl` should be redundant _anyway_: `ServerSocket.accept` will call `Socket.init(socket:)` which will call `BaseSocket.init(socket:)` which will call `self.ignoreSIGPIPE`. So we seem to be calling this fcntl twice per accepted socket. The output of `dtruss` agrees with this assessment:

```
accept(0x16, 0x0, 0x0)		 = 23 0
fcntl(0x17, 0x49, 0x1)		 = 0 0
fcntl(0x17, 0x49, 0x1)		 = 0 0
fcntl(0x17, 0x3, 0x0)		 = 6 0
fcntl(0x17, 0x4, 0x6)		 = 0 0
```

On FD 23 we are calling `fcntl` with `0x49` (`F_SETNOSIGPIPE`) twice in a row, both times setting to 1. The other two calls are associated with getting and then setting the flags for `O_NONBLOCK` (which ironically didn't need to be set in this case anyway).

So, if either of those calls to `F_SETNOSIGPIPE` throws they will throw an error that ends up terminating the channel.

I think the fix is in two parts:

1. Remove the call to `F_SETNOSIGPIPE` in `Posix.accept`. It's redundant with the `Socket` initializer anyway.
2. Update `BaseSocketProtocol.ignoreSIGPIPE(descriptor:)` to have similar logic to `Posix.setNonBlocking`: check the return code and throw a custom error that we can ignore in `shouldCloseOnReadError`.

**Comment by Lukasa** [comment]
@cyberdelia Do you feel equipped to try to fix this?

**Comment by Lukasa** [comment]
Note that we'll need a test. @weissi wrote one for the similar issue in bc81605.

**Comment by cyberdelia** [comment]
@lukasa Seems doable 🤔

**Comment by cyberdelia** [comment]
@Lukasa Gave it a shot, but looks like there is other behaviour expecting to receive an proper EINVAL: https://github.com/apple/swift-nio/blob/master/Tests/NIOTests/SocketChannelTest.swift#L596-L599

**Comment by Lukasa** [comment]
That test is overly specific, so we can change it. What matters is that we must _detect_ that the client sent a RST packet and tell the user by closing the accepted channel.

**Comment by cyberdelia** [comment]
Adapted the failing test, I'm not completely sold on the name `NIOFailedToSetSocketCommand` either.

File: Sources/NIO/System.swift [hunk 1]
```diff
@@ -574,12 +562,11 @@ internal enum Posix {
...
 }
 
-/// `NIOFailedToSetSocketNonBlockingError` indicates that NIO was unable to set a socket to non-blocking mode, either
-/// when connecting a socket as a client or when accepting a socket as a server.
+/// `NIOFailedToSetSocketCommand` indicates that NIO was unable to perform an
+/// operation on a socket.
 ///
-/// This error should never happen because a socket should always be able to be set to non-blocking mode. Unfortunately,
-/// we have seen this happen on Darwin.
-public struct NIOFailedToSetSocketNonBlockingError: Error {}
```
Comment (Lukasa) [inline Sources/NIO/System.swift:582]: We can't rename this error, it's a semver major. We should just add the new one instead. We can deprecate the `NIOFailedToSetSocketNonBlockingError` if we want to.

As for the new name, maybe `NIOFcntlFailedError`. Seems the most descriptive.

Comment (cyberdelia) [reply Sources/NIO/System.swift:582]: Went the deprecation route.

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

**Pull request merged**